### ISSUE TITLE
Cadence doesn't autostart Jack on logon despite the checkbox being checked to do so - Arch Linux 64 bit

### ISSUE THREAD

**Issue by deadbeatguitarist** [falkTX/Cadence#91]
I can start Cadence manually however if I checkbox the autostart option on Jack and the Pulse audio bridge and log out and back in in either KDE or MATE - (haven't tried others as well) it won't autostart Jack.

**Comment by milkmiruku** [comment]
I'm getting this consistently on Arch 64-bit. Adding "jack_control start &" back to my .xinitrc fixes the first half but the problem remains for the PA bridge. Is there anything I could further report to assist?

**Comment by falkTX** [comment]
Sure, see if this works instead:
´´´
cadence-session-start --system-start
´´´

That's the initial command used to start jack & bridges.

**Comment by prikhi** [comment]
`cadence-session-start --system-start` works great in my AwesomeWM config file

**Comment by falkTX** [comment]
This might now be fixed with the latest commit, please update and let me know if it works.

**Comment by rghvdberg** [comment]
ubuntu 14.04.2 

cadence / jack doesn't autostart for me neither
cadence-session-start works ok

>cadence-session-start --system-start
>cadence-aloop-daemon started, using alsa_in/out and 2 channels
>Initiated PulseAudio successfully!
>JACK Started Successfully

cadence packages installed :

>dpkg --list | grep cadence
>ii  cadence                                               1:0.8.1+git20150418.2                              all          JACK audio toolbox
>ii  cadence-data                                          1:0.8.1+git20150418.2                               all          JACK audio toolbox (data files)
>ii  cadence-tools                                         1:0.8.1+git20150418.2                               amd64        JACK audio toolbox (small tools)
>ii  cadence-unity-support                                 1:0.8.1+git20150418.2                               all          JACK audio toolbox (Unity support)

**Comment by troathscream** [comment]
Same here on Ubuntu Studio 14.04.2 with the KXstudio Repo's enabled.
I also have the problem that claudia keeps crashing cadence/jack and the bridges when making a
studio.

**Comment by prikhi** [comment]
I built & installed from master on Arch Linux using the Awesome window manager & LightDM display manager. 

On login, Claudia was started & sitting in my task bar but I had to manually start Cadence & then the JACK server to get my default studio loaded.

**Comment by prikhi** [comment]
Actually sorry, Claudia starting was caused by a line in my window manager config.

**Comment by falkTX** [comment]
AwesomeWM doesn't support X11 rules (ie, reading the files in /etc/X11/xinit/xinitrc.d/).
We'll need something that works for those non-compliant desktops but that doesn't conflict with what is already in place.

@troathscream which desktop are you using?
And do you have all the latest updates?

**Comment by troathscream** [comment]
I am using the standard WM of ubuntu studio 14.04 XFCE,
yes i have run all updates.

**Comment by troathscream** [comment]
PS it is ubuntu studio 14.04 64 Bit

**Comment by falkTX** [comment]
do a purge of cadence just in case:
```
sudo apt-get purge cadence
sudo apt-get install cadence
```

**Comment by troathscream** [comment]
i tried purging and reinstalling cadence, but no luck.
It now starts at boot but doesn't autostart the jack server and still doesn't open my claudia/ladish studio.

**Comment by falkTX** [comment]
you can add:
```
cadence-session-start --system-start
```
as a startup script (somewhere in your Desktop settings).

**Comment by prikhi** [comment]
> We'll need something that works for those non-compliant desktops but that doesn't conflict with what is already in place.

I think it's fine to just use `/etc/X11/xinit/xinitrc.d/`. I did a little digging & the recommended way to handle this with non-compliant WMs is to add something like this to your `~/.xinitrc`(I actually already had this in mine):
```bash
# Source system config
if [ -d /etc/X11/xinit/xinitrc.d ]; then
  for f in /etc/X11/xinit/xinitrc.d/*; do
    [ -x "$f" ] && . "$f"
  done
  unset f
fi
```

Then I just had to mark the files as executable:
```bash
chmod +x /etc/X11/xinit/xinitrc.d/*cadence*
```

Should this be added to the docs somewhere?

**Comment by falkTX** [comment]
that's something you need to add to the respective non-compliant WMs.
please fill a bug report against the WM you use that doesn't support this.

also, you don't need to make them executable.
the files are not executed, they are "sourced".
ie, '. /some/file.sh' vs '/some/file.sh'
your '[ -x "$f" ]' test should be '[ -f "$f" ]'

**Comment by prikhi** [comment]
> that's something you need to add to the respective non-compliant WMs.
please fill a bug report against the WM you use that doesn't support this.

Just checking, wasn't sure if there was some tucked away spot in the docs for edge cases like this

>also, you don't need to make them executable.
the files are not executed, they are "sourced".

Ah ok, I just grabbed that from the code from the Arch Wiki: https://wiki.archlinux.org/index.php/Xinitrc#Configuration
Everything else in `xinitrc.d` was marked executable so I was led to believe that was legitimate.

**Comment by brunoruviaro** [comment]
Running into the same problem here. Ubuntu 14.04 64 bit with KXStudio repositories & all recent updates made. Cadence starts OK if done manually from GUI or command line (`cadence-session-start --system-start`). Upon rebooting the computer (or complete shut down & turn back on), Jack does not autostart even though "auto start" checkboxes for Jack, MIDI, and PA were previously selected in Cadence. Currently installed versions (cadence, cadence-data, cadence-tools) is 2:0.8.1+git20150820. I have also tried the purge and re-install. Let me know if there is any further testing I can do to help find the solution...

**Comment by kmchen13** [comment]
Same on Debian Jessie when I resume from suspend ( sound works fine after normal boot):
`kmc@kmcs:~$ cadence-session-start -s
Initiated PulseAudio successfully!
JACK Started Successfully
kmc@kmcs:~$ cadence-session-start --system-start
Initiated PulseAudio successfully!
JACK Started Successfully
kmc@kmcs:~$ ^C
kmc@kmcs:~$ dpkg --list | grep cadence                                                   
ii  cadence                                2:0.8.1+git20151026                  all          JACK audio toolbox                                                                   
ii  cadence-data                           2:0.8.1+git20151026                  all          JACK audio toolbox (data files)
ii  cadence-tools                          2:0.8.1+git20151026                  amd64        JACK audio toolbox (small tools)
`
Here are the logs:
[
Apr 12 08:48:45 kmcs org.jackaudio.service[22449]: no message buffer overruns
Apr 12 08:48:45 kmcs org.jackaudio.service[22449]: no message buffer overruns
Apr 12 08:48:45 kmcs rtkit-daemon[1740]: Successfully made thread 20486 of process 20486 (/usr/bin/pulseaudio) owned by '1000' high priority at nice level -11.
Apr 12 08:48:45 kmcs rtkit-daemon[1740]: Supervising 1 threads of 1 processes of 1 users.
Apr 12 08:48:45 kmcs pulseaudio[20486]: Stale PID file, overwriting.
Apr 12 08:48:45 kmcs org.jackaudio.service[22449]: no message buffer overruns
Apr 12 08:48:45 kmcs pulseaudio[20486]: JACK error >Cannot connect to server socket err = Connexion refusée<
Apr 12 08:48:45 kmcs pulseaudio[20486]: JACK error >Cannot connect to server request channel<
Apr 12 08:48:46 kmcs pulseaudio[20486]: Default and alternate sample rates are the same.
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >Cannot use real-time scheduling (RR/5)(1: Opération non permise)<
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >JackClient::AcquireSelfRealTime error<
Apr 12 08:48:46 kmcs pulseaudio[20486]: Default and alternate sample rates are the same.
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >Cannot use real-time scheduling (RR/5)(1: Opération non permise)<
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >JackClient::AcquireSelfRealTime error<
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >Server is not running<
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >Server is not running<
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >Server is not running<
Apr 12 08:48:46 kmcs pulseaudio[20486]: JACK error >Server is not running<
Apr 12 08:48:47 kmcs org.jackaudio.service[22449]: no message buffer overruns
Apr 12 08:48:47 kmcs org.jackaudio.service[22449]: no message buffer overruns
Apr 12 08:48:47 kmcs org.jackaudio.service[22449]: no message buffer overruns
Apr 12 08:48:47 kmcs org.jackaudio.service[22449]: ALSA lib pcm_hw.c:1667:(_snd_pcm_hw_open) Invalid value for card
Apr 12 08:48:47 kmcs rtkit-daemon[1740]: Successfully made thread 20567 of process 20567 (/usr/bin/pulseaudio) owned by '1000' high priority at nice level -11.
Apr 12 08:48:47 kmcs rtkit-daemon[1740]: Supervising 1 threads of 1 processes of 1 users.
Apr 12 08:48:47 kmcs pulseaudio[20567]: Stale PID file, overwriting.
Apr 12 08:48:47 kmcs pulseaudio[20567]: Default and alternate sample rates are the same.
Apr 12 08:48:47 kmcs pulseaudio[20567]: JACK error >Cannot use real-time scheduling (RR/5)(1: Opération non permise)<
Apr 12 08:48:47 kmcs pulseaudio[20567]: JACK error >JackClient::AcquireSelfRealTime error<
Apr 12 08:48:47 kmcs pulseaudio[20567]: Not enough physical output ports, leaving unconnected.
Apr 12 08:48:48 kmcs pulseaudio[20567]: Default and alternate sample rates are the same.
Apr 12 08:48:48 kmcs pulseaudio[20567]: JACK error >Cannot use real-time scheduling (RR/5)(1: Opération non permise)<
Apr 12 08:48:48 kmcs pulseaudio[20567]: JACK error >JackClient::AcquireSelfRealTime error<
](url)

**Comment by juliandroid** [comment]
I have the same issue.
In Arch Linux, /etc/X11/xinit/xinitrc looks partially like this:

```
# start some nice programs

if [ -d /etc/X11/xinit/xinitrc.d ] ; then
 for f in /etc/X11/xinit/xinitrc.d/?*.sh ; do
  [ -x "$f" ] && . "$f"
 done
 unset f
fi

```
The generated files has to be executable, otherwise there are not started/loaded:
70cadence-plugin-paths
99cadence-session-start

The other files in the same directory looks like this:

> -rwxr-xr-x 1 root root  318 May  5 00:59 40-libcanberra-gtk-module.sh
> -rwxr-xr-x 1 root root  198 Jan 29 00:32 50-systemd-user.sh

**Comment by ericfont** [comment]
I believe I have had the same or similar problem described on Arch Linux Mate.  I had Cadence added to my Mate autostart.  However, Cadence would fail to start jack because 'defualt' was already being utilized. 
  I could however start jack via the "Force Restart" button fine.  I believe that mate desktop was the culprit which had taken over my audio device, preventing a non-force start of jack.

So my solution was to start non-gui cadence via .xinitrc before starting the mate desktop:
```
cadence-session-start --system-start
exec mate-session
```

I found that did cause the desktop session to take a longer time to start, however, presumably because jack is getting setup.  Anyway, just letting people know how it I got it working.

**Comment by juliandroid** [comment]
As I've said above, looking at the xinitrc, having cadence scripts with execute bit set is a must (i.e. will not run), so my fix is setting those scripts with +x and those are now working under Arch linux.

**Comment by falkTX** [comment]
If it's just the executable bit, then I guess I can just change it now.
Please test.

**Comment by SicariusNoctis** [comment]
The executable bit doesn't seem to be set on a recently installed copy of cadence:

![Executable bit unset](https://i.imgur.com/QLtg5bE.png)

**Comment by falkTX** [comment]
You seem to have an old version.
Current makefile installs a single file to xinit, see https://github.com/falkTX/Cadence/blob/e91750e39acb119cf97f8105a7acc28a8901a4e2/Makefile#L167

it even tries to remove files you have installed there, no longer in use, see https://github.com/falkTX/Cadence/blob/e91750e39acb119cf97f8105a7acc28a8901a4e2/Makefile#L189

**Comment by SicariusNoctis** [comment]
Ah, I installed from [AUR](https://aur.archlinux.org/packages/cadence/). It seems to reference [SourceForge binaries](https://sourceforge.net/projects/kxstudio/files/Releases/cadence/) from 2013.

**Comment by milkmiruku** [comment]
@SicariusNoctis; to be sure, [cadence-git](https://aur.archlinux.org/packages/cadence-git) is also in the AUR.

**Comment by redtide** [comment]
[Cadence 0.9.0](https://www.archlinux.org/packages/community/x86_64/cadence/) is now in the community repository.

I don't understand why sometimes I get sound and sometimes not but I notice that on /etc/X11/xinit/xinitrc the line that starts the application is
```
...
for f in /etc/X11/xinit/xinitrc.d/?*.sh ; do
...
```
so are only .sh files be executed? All files inside the initrc.d are in the form of NN-some-name.sh, so should it be named also 61-cadence-session-inject.sh instead?

**Comment by xxRockOnxx** [comment]
> [Cadence 0.9.0](https://www.archlinux.org/packages/community/x86_64/cadence/) is now in the community repository.
> 
> I don't understand why sometimes I get sound and sometimes not but I notice that on /etc/X11/xinit/xinitrc the line that starts scripts is
> 
> ```
> ...
> for f in /etc/X11/xinit/xinitrc.d/?*.sh ; do
> ...
> ```
> 
> so are only .sh files to be executed? All files inside the initrc.d directory are in the form of NN-some-name.sh, so should it be named 61-cadence-session-inject.sh instead?

I'm on Arch and have installed Cadence 0.9.1-2 and this still seems to be an issue.

**Comment by Marzal** [comment]
ArchLinux KDE
Cadence 0.9.1-2
No problem with autostart.

I don't have any `/etc/X11/xinit/xinitrc` file (no xorg-xinit package installed). What I have is the **/usr/share/sddm/scripts/Xsession** file that sources every file in xinitrc.d
```bash
# run all system xinitrc shell scripts.
if [ -d /etc/X11/xinit/xinitrc.d ]; then
  for i in /etc/X11/xinit/xinitrc.d/* ; do
  if [ -x "$i" ]; then
    . "$i"
  fi
  done
fi
```
The variables are exported but I think that STARTUP is never executed (at least I can't find any file in my system using that)
`STARTUP="$INSTALL_PREFIX/bin/cadence-session-start --system-start-by-x11-startup $STARTUP"`

But **/etc/xdg/autostart/[cadence-session-start.desktop](https://github.com/falkTX/Cadence/blob/master/data/autostart/cadence-session-start.desktop)** launch `cadence-session-start --maybe-system-start` so everything works ok.

**Comment by YodaEmbedding** [comment]
I think for Arch and other systemd based distros, it would make sense to have a systemd startup service take care of running pulseaudio and jack _in the correct order_. Currently, I am just hackily starting things 20 seconds after startup to make sure things are loaded properly.

**Comment by TheBeachLab** [comment]
> I think for Arch and other systemd based distros, it would make sense to have a systemd startup service take care of running pulseaudio and jack _in the correct order_. Currently, I am just hackily starting things 20 seconds after startup to make sure things are loaded properly. Systemd provides more control by starting services in some deterministic order.

Master Yoda, could you please share your approach?

**Comment by TheBeachLab** [comment]
> The variables are exported but I think that STARTUP is never executed (at least I can't find any file in my system using that)
> `STARTUP="$INSTALL_PREFIX/bin/cadence-session-start --system-start-by-x11-startup $STARTUP"`

I am relatively new to Linux but I think there's a mistake in this line. It looks like @falkTX is calling $STARTUP from inside STARTUP itself. This solved the problem in Arch Linux with i3 wm:

```bash
STARTUP="$INSTALL_PREFIX/bin/cadence-session-start --system-start-by-x11-startup"
$STARTUP
```

Now the behaviour is correct as expected

**Comment by Marzal** [comment]
I thought it was appending that content to the variable STARTUP, leaving up to Xsession / xinit to execute it.
http://xpt.sourceforge.net/techdocs/nix/x/general/xwin03-WinManagerdesktopsChoosing/ar01s04.html

**Comment by TheBeachLab** [comment]
Thanks @Marzal , I get it now. My wm seems not to use this variable, that's why it didn't autostart if not called explicitly

**Comment by redtide** [comment]
> I think for Arch and other systemd based distros, it would make sense to have a systemd startup service take care of running pulseaudio and jack _in the correct order_. Currently, I am just hackily starting things 20 seconds after startup to make sure things are loaded properly. Systemd provides more control by starting services in some deterministic order.

That's what I ended up doing on my own application, for [JACK](https://github.com/jackaudio/jack2/tree/develop/systemd) at least (I'm not using PA).

**Comment by tatrapikao** [comment]
Hmm - On Xubuntu 20.10 "cadence-session-start --maybe-system-start" in autostart will never succeed. But when i start cadence manually without starting Jack i can run "cadence-session-start --maybe-system-start" and it will succeed starting Jack and bridging pulse. Something like  "cadence && sleep 3 && cadence-session-start --maybe-system-start" works sometimes but never from autostart.  On a different System also with Xubuntu 20.10 most of the time autostart works fine out of the box (not always though).

**Comment by lgmventura** [comment]
I had this problem when I first installed Jack. After reinstalling my Ubuntu and everything except the `home` folder, this problem was gone. Now, this problem is suddenly back. I'm currently using Ubuntu 20.04.

**Comment by bilbodwyer** [comment]
FWIW I was having this issue, and amending the file at /etc/X11/xinit/xinitrc.d/61cadence-session-inject to read 
```
STARTUP="$INSTALL_PREFIX/bin/cadence-session-start --system-start-by-x11-startup" 
$STARTUP
```

did the trick in getting jackd to autostart correctly. However, I was coming across a separate issue whereby Firefox and Reaper could play audio, but Spotify and GuitarPro (closed-source applications, granted) could not. Changing the STARTUP variable to read
``STARTUP="$INSTALL_PREFIX/bin/cadence-session-start -s"
solved that issue as well.

**Comment by GabubuAvailable** [comment]
Heyo o/
I tried the "cadence-session-start --system-start" command too but I receive this error:
I use Arch Linux with the latest version of Cadence from the [community repository](https://archlinux.org/packages/community/x86_64/cadence/)
![output](https://user-images.githubusercontent.com/88589756/158035878-529af8b4-da0b-4a9d-9f9f-2c571d029aa9.png)
